Source/Atomic/Network/MasterServerClient.h

```diff
@@ -0,0 +1,128 @@
+//
+// Communication with an atomic master server
+// Handles NAT Punch-through for clients
+//
+
+#pragma once
+
+#include <ThirdParty/kNet/include/kNet.h>
+#include "../Core/Object.h"
+
+namespace Atomic
+{
+
+class Scene;
+
+enum ConnectToMasterState
+{
+    MASTER_NOT_CONNECTED = 0,
+    MASTER_CONNECTING_UDP,
+    MASTER_CONNECTING_TCP,
+    MASTER_CONNECTED,
+    MASTER_CONNECTION_FAILED
+};
+
+enum ClientConnectToGameServerState
+{
+    GAME_NOT_CONNECTED = 0,
+    GAME_CONNECTING_INTERNAL_IP,
+    GAME_VERIFYING_SERVER,
+    GAME_CONNECTING_EXTERNAL_IP,
+    GAME_CONNECTED,
+    GAME_CONNECTION_FAILED
+};
+
+struct MasterServer
+{
+    String address;
+    unsigned short port;
+    bool isRegisteringServer;
+    String serverName;
+};
+
+/// Game server info used by a client when connecting
+struct RemoteGameServer
+{
+    /// Game server id
+    String serverId;
+    String internalAddress;
+    unsigned short internalPort;
+    String externalAddress;
+    unsigned short externalPort;
+    Scene* clientScene;
```

Comment:
Any chance of this being released while connecting?  If so, should it be a SharedPtr or WeakPtr depending on whether a strong reference is required?